core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -282,30 +284,36 @@ private ProjectNode createPreProjection(
     {
-        Set<PreAggregationKey> keys = new HashSet<>();
-        ImmutableMap.Builder<PreAggregationKey, PreAggregation> preAggregations = ImmutableMap.builder();
-        for (CaseAggregation aggregation : aggregations) {
-            PreAggregationKey preAggregationKey = new PreAggregationKey(aggregation);
-            if (keys.contains(preAggregationKey)) {
-                continue;
-            }
-
-            // Cast pre-projection if needed to match aggregation input type.
-            // This is because entire "CASE WHEN" expression could be wrapped in CAST.
-            Expression preProjection = aggregation.getResult();
-            Type preProjectionType = getType(context, preProjection);
-            Type aggregationInputType = getOnlyElement(aggregation.getFunction().getSignature().getArgumentTypes());
-            if (!preProjectionType.equals(aggregationInputType)) {
-                preProjection = new Cast(preProjection, toSqlType(aggregationInputType));
-                preProjectionType = aggregationInputType;
-            }
-
-            Symbol preProjectionSymbol = context.getSymbolAllocator().newSymbol(preProjection, preProjectionType);
-            Symbol preAggregationSymbol = context.getSymbolAllocator().newSymbol(aggregation.getAggregationSymbol());
-            preAggregations.put(preAggregationKey, new PreAggregation(preAggregationSymbol, preProjection, preProjectionSymbol));
-            keys.add(preAggregationKey);
-        }
-        return ImmutableMap.copyOf(preAggregations.buildOrThrow());
+        return aggregations.stream()
+                .collect(toImmutableSetMultimap(PreAggregationKey::new, identity()))
+                .asMap().entrySet().stream().collect(toImmutableMap(
+                        Map.Entry::getKey,
+                        entry -> {
+                            PreAggregationKey key = entry.getKey();
+                            Set<CaseAggregation> caseAggregations = (Set<CaseAggregation>) entry.getValue();
+                            Expression preProjection = key.projection;
+
+                            // Cast pre-projection if needed to match aggregation input type.
+                            // This is because entire "CASE WHEN" expression could be wrapped in CAST.
+                            Type preProjectionType = getType(context, preProjection);
+                            Type aggregationInputType = getOnlyElement(key.getFunction().getSignature().getArgumentTypes());
+                            if (!preProjectionType.equals(aggregationInputType)) {
+                                preProjection = new Cast(preProjection, toSqlType(aggregationInputType));
+                                preProjectionType = aggregationInputType;
+                            }
+
+                            // Wrap the preProjection with IF to retain the conditional nature on the CASE aggregation(s) during pre-aggregation
```

Comment:
nit: `...conditional nature on the CASE aggregation(s) during pre-aggregation` . It's more than just that, e.g: divison by 0